app/services/google/parser_service.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module Google
+  class ParserService
+    NON_ADS_RESULT_SELECTOR = 'a[data-ved]:not([role]):not([jsaction]):not(.adwords):not(.footer-links)'
+    AD_CONTAINER_ID = 'tads'
+    ADWORDS_CLASS = 'adwords'
+
+    def initialize(html_response:)
+      raise ArgumentError, 'response.body cannot be blank' if html_response.body.blank?
+
+      @html = html_response
+
+      @document = Nokogiri::HTML.parse(html_response)
+
+      # Add a class to all AdWords link for easier manipulation
+      document.css('div[data-text-ad] a[data-ved]').add_class(ADWORDS_CLASS)
+
+      # Mark footer links to identify them
+      document.css('#footcnt a').add_class('footer-links')
+    end
+
+    # Parse html data and return a hash with the results
+    def call
```

Comment:
And then in this method, we will early return if it does not pass the `valid?` check.

```suggestion
    def call
      return unless valid?

```

```ruby
def valid?
  html.present? && document.present?
end
```

Reply:
7dda9bb68f77101383b233bece3f91aa5b25f3a7. Did a bit of refactoring based on your suggestions. Please let me know if this covers everything.

Reply:
Please sort the order of methods. We usually put `valid?, mark_adword_links, mark_footer_links, present_parsed_data` on top of other ones.

Reply:
Fixed in bd1df63230d99ee4c173bac6d94bb834bb6eec43